### PR TITLE
CASEFLOW-376 (Part 1) Allow optional dates for straight_reversal and death_dismissal

### PR DESCRIPTION
**Pull request by yoomlam**
Connects #15820 (Part 1 of [CASEFLOW-376](https://vajira.max.gov/browse/CASEFLOW-376))

Reason: So that CASEFLOW-376 can be demo'd.

### Description
For straight_reversal and death_dismissal types of CAVC Remands, allow judgement & mandate dates to be optional.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] For straight_reversal and death_dismissal, judgement & mandate dates are optional when filling out the form.

### Testing Plan
(See Testing Plan in #15774 for more detailed instructions.)
1. Fill out a CAVC Remand form, picking either straight_reversal or death_dismissal.
2. Submit form without judgement & mandate dates
3. Check that resulting appeal is created

### CODE REVIEW

File: app/workflows/initial_tasks_factory.rb [hunk 1]
```diff
@@ -54,12 +54,18 @@ def create_subtasks!
...
+    if @cavc_remand.remand?
+      if @cavc_remand.mdr?
+        MdrTask.create_with_hold(cavc_task)
+      elsif @cavc_remand.jmr? || @cavc_remand.jmpr?
+        SendCavcRemandProcessedLetterTask.create!(appeal: @appeal, parent: cavc_task)
+      else
+        fail "Not yet supported remand subtype: #{@cavc_remand.remand_subtype}"
+      end
+    elsif @cavc_remand.straight_reversal? || @cavc_remand.death_dismissal?
+      puts "TBD======"
```
Comment (yoomlam) [inline app/workflows/initial_tasks_factory.rb:66]: Will be completed in #15774.

File: app/models/cavc_remand.rb [hunk 2]
```diff
@@ -17,9 +17,13 @@ class CavcRemand < CaseflowRecord
...
   validates :cavc_judge_full_name, inclusion: { in: Constants::CAVC_JUDGE_FULL_NAMES }
   validates :remand_subtype, presence: true, if: :remand?
-  validates :judgement_date, :mandate_date, presence: true, unless: -> { remand? && mdr? }
+  validates :judgement_date, :mandate_date, presence: true, unless: :mandate_not_required?
   validate :decision_issue_ids_match_appeal_decision_issues, if: -> { remand? && jmr? }
 
+  def mandate_not_required?
+    straight_reversal? || death_dismissal? || (remand? && mdr?)
+  end
+
```
Comment (yoomlam) [inline app/models/cavc_remand.rb:26]: I usually try to create positive methods (`mandate_required?`) but this implementation (`!(straight_reversal? || death_dismissal? || (remand? && mdr?))`) required more mental power.

File: app/controllers/cavc_remands_controller.rb [hunk 3]
```diff
@@ -54,15 +54,21 @@ def validate_cavc_remand_access
...
+  def required_params_by_decisiontype_and_subtype
+    case params["cavc_decision_type"]
+    when Constants.CAVC_DECISION_TYPES.remand
+      case params["remand_subtype"]
+      when Constants.CAVC_REMAND_SUBTYPES.mdr
+        REMAND_REQUIRED_PARAMS
+      else
+        REMAND_REQUIRED_PARAMS + JMR_REQUIRED_PARAMS
+      end
+    when Constants.CAVC_DECISION_TYPES.straight_reversal, Constants.CAVC_DECISION_TYPES.death_dismissal
```
Comment (ajspotts) [inline app/controllers/cavc_remands_controller.rb:70]: first time I'm seeing the usage of `case` 🤩 , reminds me of [`switch`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch) in JS

File: app/models/cavc_remand.rb (hunk 2 above)
Comment (ajspotts) [reply app/models/cavc_remand.rb:26]: agreed - the combo of negatives & positives within a conditional can get ugly

File: client/app/queue/AddCavcRemandView.jsx [hunk 4]
```diff
@@ -113,12 +113,15 @@ const AddCavcRemandView = (props) => {
     setIssues({ ...issues, [evt.target.name]: evt.target.checked });
   };
 
+  const straightReveralType = () => type === CAVC_DECISION_TYPES.straight_reversal;
```
Comment (ajspotts) [inline client/app/queue/AddCavcRemandView.jsx:116]: typo - think you want straightReversalType (missing the S)

File: app/workflows/initial_tasks_factory.rb [hunk 5]
```diff
@@ -54,12 +54,21 @@ def create_subtasks!
...
-      MdrTask.create_with_hold(cavc_task)
-    elsif @cavc_remand.jmr? || @cavc_remand.jmpr?
-      SendCavcRemandProcessedLetterTask.create!(appeal: @appeal, parent: cavc_task)
+    case @cavc_remand.cavc_decision_type
+    when Constants.CAVC_DECISION_TYPES.remand
+      case @cavc_remand.remand_subtype
+      when Constants.CAVC_REMAND_SUBTYPES.mdr
+        MdrTask.create_with_hold(cavc_task)
+      when Constants.CAVC_REMAND_SUBTYPES.jmr, Constants.CAVC_REMAND_SUBTYPES.jmpr
+        SendCavcRemandProcessedLetterTask.create!(appeal: @appeal, parent: cavc_task)
```
Comment (ajspotts) [inline app/workflows/initial_tasks_factory.rb:63]: out of curiosity, could you have used `@cavc_remand.mdr?` in place of Constants.CAVC_REMAND_SUBTYPES.mdr (same question for `jmr`) or no?

Comment (yoomlam) [reply app/workflows/initial_tasks_factory.rb:63]: Not when using the `case/when` block, which expects constants or regex: https://www.rubyguides.com/2015/10/ruby-case/. `mdr?` is a method.

Comment (ajspotts) [reply app/workflows/initial_tasks_factory.rb:63]: got it thanks for clarifying

File: app/controllers/cavc_remands_controller.rb (hunk 3 above)
Comment (yoomlam) [reply app/controllers/cavc_remands_controller.rb:70]: Yup, I learned about `case/when` yesterday. 😆

**Review by ajspotts — APPROVED** [review]
ran through UI testing w/ successful creation of SR & DD CavcRemands sans judgement & mandate dates. lgtm 🚀

**Pull request merged**